### ISSUE TITLE
Cannot build react component with hooks to IIFE

### ISSUE THREAD

**Issue by Wonkers0** [evanw/esbuild#3251]
I want to bundle a react component to an IIFE that I can then serve over HTTP and `eval()` on the client.

Kind of like this:
```tsx
const compiledCode = `...`;
const TestComponent = eval(compiledCode).default;

function App() {
  return <TestComponent />;
}

export default App;
```

[Here's the TestComponent source](https://esbuild.github.io/try/#YgAwLjE4LjE0AHsKICBidW5kbGU6IHRydWUsCgogIHBsYXRmb3JtOiAiYnJvd3NlciIsCiAgZm9ybWF0OiAiaWlmZSIsCiAgZ2xvYmFsTmFtZTogIlRlc3RDb21wb25lbnQiLAogIGZvb3RlcjogewogICAganM6ICJUZXN0Q29tcG9uZW50OyIsIC8vIFRoaXMgaXMgdGhlIGdsb2JhbCB2YXJpYWJsZSBuYW1lCiAgfSwKICBsb2FkZXI6IHsnLmpzJzogJ2pzeCd9LAogIGV4dGVybmFsOiBbInJlYWN0Il0KfQBlAGVudHJ5LmpzAGltcG9ydCB7IHVzZUVmZmVjdCB9IGZyb20gInJlYWN0IjsKCmV4cG9ydCBkZWZhdWx0IGZ1bmN0aW9uIFRlc3RDb21wb25lbnQoKSB7CiAgdXNlRWZmZWN0KCgpID0+IHsKICAgIGNvbnNvbGUubG9nKCJNb3VudGVkIik7CiAgfSk7CgogIHJldHVybiA8YnV0dG9uPlRlc3QgQ29tcG9uZW50PC9idXR0b24+Owp9Cg)

If I mark "react" as external: 
![image](https://github.com/evanw/esbuild/assets/106038003/88b0d21d-fbfc-4bc0-aba3-764b12aa296d)

If I don't: (Keep in mind that I run the build producing the errors below with `minify: true`
![image](https://github.com/evanw/esbuild/assets/106038003/b1247359-6c11-44f5-90e1-a8fa4aec6642)

I'm by no means an expert on web bundlers, I just want to get this done so I can continue working on my project, which is why I'm looking for guidance, so any help would be appreciated.

**Comment by evanw** [comment]
Sorry, direct `eval` is not supported: https://esbuild.github.io/content-types/#direct-eval. You could consider using ES modules and a [dynamic `import`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) instead. Another option is using a dynamically-generated [`<script>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script).

**Comment by Wonkers0** [comment]
> Sorry, direct `eval` is not supported: https://esbuild.github.io/content-types/#direct-eval. You could consider using ES modules and a [dynamic `import`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) instead. Another option is using a dynamically-generated [`<script>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script).

Hey, thanks for the help. I tried dynamic imports, like so:
```ts
async function importModule(moduleCode: string) {
  try {
    const blob = new Blob([moduleCode], { type: "application/javascript" });

    const url = URL.createObjectURL(blob);
    return await import(url);
  } catch (error) {
    console.error("Error importing the module:", error);
  }
}
```

However I'm getting this error if I mark the imported libraries as external:
![image](https://github.com/evanw/esbuild/assets/106038003/cc77c334-1122-493c-a82c-c2c229079b49)

And this error if I don't (with `bundle: true`):
![image](https://github.com/evanw/esbuild/assets/106038003/205fa031-6202-423c-8948-9488510734dc)
![image](https://github.com/evanw/esbuild/assets/106038003/f47f08af-7279-4153-bf63-a7373ba28060)

As far as dynamically-generated <script> tags go, I'm not sure how I would import anything from an inline module, since it doesn't have a file name.

**Comment by hyrious** [comment]
What you've done is not a typical code splitting practice. Generally we just use dynamic imports and enable the code [splitting](https://esbuild.github.io/api/#splitting) and it is done (the component will be downloaded when needed).

```js
// possible code
import {Suspense, lazy} from 'react'

const Test = lazy(() => import('./Test'))

const App = () => {
  return <Suspense><Test /></Suspense>
}
```

**Comment by Wonkers0** [comment]
I'm not sure I understand. In the sample you've given:
```js
const Test = lazy(() => import('./Test'))
```

It's importing a module from a separate **file**, but I'm trying to send the module code **over HTTP** so that I can first authenticate the client, and only then serve the component. So all I have on the client is just a string of the bundled code, but I'm not sure how I'd get the default export from that. (I tried to `eval()` an IIFE initially)

**Comment by hyrious** [comment]
When you call `import()` at front-end (i.e. in a browser), it then becomes a `fetch()` call to that file, you can do whatever you like when the server got that request, e.g. authenticaton.

**Comment by hyrious** [comment]
I'd like to avoid using `eval()`-like things in production, unless necessary. However if you insist on that, here are some possible ways:

1. In your first example you have built the IIFE bundle with `external: ['react']`. You will get some code like this:

   ```js
   (() => {
     var __require = ... // delegates the global 'require'
     var react = __require("react")
     var TestComponent = () => {
       const ref = react.useRef()
   // ...
   ```

   As you can see, the code will try to get the 'react' module with a calling of `require('react')`. This gives you a way to make it work:

   ```js
   var require = (mod) => {
     if (mod == 'react') return React;
   }
   eval(moduleCode)
   ```

   Another thing is the IIFE does not return any thing. You can enable `--global-name=X` to assign the module export to a variable (and then return it at the end of the code):

   ```js
   // compiled module with --global-name=X:
   var moduleCode = 'var X = (() => { ... })()'
   // 'eval(moduleCode)' change to:
   var mod = new Function(moduleCode + '\n return X')()
   ```

2. Can `import()` be used to evaluate the string? Yes. It is already covered in https://2ality.com/2019/10/eval-via-import.html. It has some problems like its stack trace is not a normal filename but a random hash or long string. Besides, you can make use of [importmap](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap) to make imports of external modules in the string work.

   ```js
   // compiled module with --format=esm
   var moduleCode = `import React from 'react'\nexport default 1`
   // evaluate via import()
   var mod = await import('data:text/javascript;charset=utf-8,' + Math.random() + ';' + encodeURIComponent(moduleCode)).then(r => r.default)
   ```

**Comment by Wonkers0** [comment]
I agree with you that `eval()` is messy. In fact, I'd prefer to simply serve a module.

[Check this demo out](https://esbuild.github.io/try/#YgAwLjE4LjE0AC0tYnVuZGxlIC0tZm9ybWF0PWVzbSAtLXNwbGl0dGluZyAtLW91dGRpcj0vIC0tcGFja2FnZXM9ZXh0ZXJuYWwAAEFwcC50c3gAaW1wb3J0IFJlYWN0LCB7IHVzZUVmZmVjdCwgdXNlU3RhdGUgfSBmcm9tICJyZWFjdCI7CmxldCBTcXVhcmUgPSBudWxsOwoKZXhwb3J0IGRlZmF1bHQgZnVuY3Rpb24gQXBwKCkgewogIGNvbnN0IFtsb2FkZWQsIHNldExvYWRlZF0gPSB1c2VTdGF0ZShmYWxzZSk7CgogIHVzZUVmZmVjdCgoKSA9PiB7CiAgICAvLyBAdHMtaWdub3JlCiAgICBpbXBvcnQoImh0dHA6Ly9sb2NhbGhvc3Q6MTUwMC8iKS50aGVuKChtb2R1bGUpID0+IHsKICAgICAgU3F1YXJlID0gbW9kdWxlLmRlZmF1bHQ7CiAgICAgIHNldExvYWRlZCh0cnVlKTsKICAgIH0pOwogIH0pOwoKICAvLyBAdHMtaWdub3JlCiAgcmV0dXJuIFNxdWFyZSA/IDxTcXVhcmUgLz4gOiA8cD5Mb2FkaW5nPC9wPjsKfQoAZQBUZXN0LnRzeABpbXBvcnQgeyBtb3Rpb24sIHVzZUFuaW1hdGlvbkNvbnRyb2xzIH0gZnJvbSAiZnJhbWVyLW1vdGlvbiI7CmltcG9ydCBSZWFjdCwgeyB1c2VFZmZlY3QgfSBmcm9tICJyZWFjdCI7CgpleHBvcnQgZGVmYXVsdCBmdW5jdGlvbiBTcXVhcmUoKSB7CiAgY29uc3QgY29udHJvbHMgPSB1c2VBbmltYXRpb25Db250cm9scygpOwoKICB1c2VFZmZlY3QoKCkgPT4gewogICAgY29uc3QgYW5pbWF0aW9uID0gYXN5bmMgKCkgPT4gewogICAgICBhd2FpdCBjb250cm9scy5zdGFydCh7IHNjYWxlOiAyLCByb3RhdGU6IDM2MCwgbGVmdDogIjUwJSIsIHRvcDogIjUwJSIgfSk7CiAgICAgIGNvbnRyb2xzLnNldCh7CiAgICAgICAgcm90YXRlOiAwLAogICAgICB9KTsKICAgICAgYXdhaXQgY29udHJvbHMuc3RhcnQoewogICAgICAgIHJvdGF0ZTogWzAsIC00NSwgNTQwXSwKICAgICAgICBib3JkZXJSYWRpdXM6IFsiNnB4IiwgIjUwJSJdLAogICAgICAgIGZpbHRlcjogWyJibHVyKDBweCkiLCAiYmx1cigxcHgpIiwgImJsdXIoMHB4KSJdLAogICAgICAgIHRyYW5zaXRpb246IHsKICAgICAgICAgIGR1cmF0aW9uOiAxLjI1LAogICAgICAgICAgZmlsdGVyOiB7CiAgICAgICAgICAgIHRpbWVzOiBbMC42LCAwLjg1LCAxXSwKICAgICAgICAgIH0sCiAgICAgICAgICBib3JkZXJSYWRpdXM6IHsKICAgICAgICAgICAgdGltZXM6IFswLjUsIDAuOTVdLAogICAgICAgICAgICBkdXJhdGlvbjogMS4yNSwKICAgICAgICAgIH0sCiAgICAgICAgICByb3RhdGU6IHsKICAgICAgICAgICAgdGltZXM6IFswLCAwLjUsIDFdLAogICAgICAgICAgICBkdXJhdGlvbjogMS4yNSwKICAgICAgICAgIH0sCiAgICAgICAgfSwKICAgICAgfSk7CiAgICB9OwoKICAgIGFuaW1hdGlvbigpOwogIH0sIFtdKTsKCiAgcmV0dXJuICgKICAgIDxtb3Rpb24uYnV0dG9uCiAgICAgIGFuaW1hdGU9e2NvbnRyb2xzfQogICAgICBjbGFzc05hbWU9InctMTIgYXNwZWN0LXNxdWFyZSBiZy1uZXV0cmFsLTMwMCByb3VuZGVkLW1kIGFic29sdXRlIgogICAgICB0cmFuc2l0aW9uPXt7CiAgICAgICAgZGVsYXk6IDEsCiAgICAgICAgZHVyYXRpb246IDIsCiAgICAgICAgdHlwZTogInNwcmluZyIsCiAgICAgIH19CiAgICA+PC9tb3Rpb24uYnV0dG9uPgogICk7Cn0K)

The module is being imported fine, but it seems that the external dependencies cannot be resolved:
![image](https://github.com/evanw/esbuild/assets/106038003/de69b28a-eb2f-424d-9f31-60427978640a)

I tried to point to the `node_modules` folder, but then it tries to fetch the dependencies from the server (do I have to expose them?):
![image](https://github.com/evanw/esbuild/assets/106038003/65bc2aeb-6c51-4798-ac12-4064caf58b4b)

I was hoping that it would try to import the dependencies from the client when the module is evaluated, but I'm not sure how that would work considering that the dependency code would already be bundled and the path probably won't line up.

Surprisingly, trying to expose the dependencies on the server still fails:
![image](https://github.com/evanw/esbuild/assets/106038003/8419e1ef-17d7-4ebb-af9c-7e89126aeb11)

(Although it seems that it can resolve "framer-motion")
I modified the bundle's import paths to point to what the path would normally resolve to: 
```js
import {
  motion as a,
  useAnimationControls as o,
} from "./node_modules/framer-motion/dist/es/index.mjs";
import r, { useEffect as e } from "./node_modules/react/index.js";
```

I feel like exposing my dependencies on the server-side is probably not the best idea, even if I can sort of hide the path with some routing. There must be a better way, though.
